Formula: cache methods' output to reduce io

xu-cheng opened the pull request:
cc @jacknagel for opinion.

jacknagel commented:
Any benchmarks? I don't think we need to cache unless it's a hotspot. And if we do, we need to make sure that no callers depend on the value reflecting changes between calls (I don't think they do).

Usually callers cache the value of e.g. `Formula.installed`.

jacknagel commented on Library/Homebrew/formula.rb:
```diff
@@ -624,22 +624,22 @@ def python(options={}, &block)
   # an array of all core {Formula} names
   def self.core_names
-    Dir["#{HOMEBREW_LIBRARY}/Formula/*.rb"].map{ |f| File.basename f, ".rb" }.sort
+    @@core_names ||= Dir["#{HOMEBREW_LIBRARY}/Formula/*.rb"].map{ |f| File.basename f, ".rb" }.sort
```
`@core_names` on the class is fine; we don't need to use class variables.

jacknagel commented on Library/Homebrew/formula.rb:
```diff
@@ -624,22 +624,22 @@ def python(options={}, &block)
   # an array of all {Formula} names, which the tap formulae have the fully-qualified name
   def self.full_names
-    core_names + tap_names
+    @@full_names ||= (core_names + tap_names)
```
parentheses are superfluous here.

xu-cheng commented:
Caching these result can significantly improve the performance on `brew audit --strict` with multiple formulae (especially when auditing all of formulae). And I haven't found caller participates the changes between calls.

I have addressed your comment and update the PR.
